### ISSUE TITLE
Timeout expired waiting for volumes to attach

### ISSUE THREAD

**Issue by hmenn** [kubernetes/kubernetes#67014]
/kind bug
/sig azure
/sig storage

**What happened**:

I'm triggering docker slaves from my jenkins master. Jenkins connectes kubernetes services and starts kubernetes pods. This pods requires some disks for persistent data(compile results actually). I'm connecting azure premium managed disk to this slaves(pods). Connecting disks takes at least 10 minutes. I'm using D4s_v3 - 4vCPU - 16Ram Virtual Machines.

**Here is my kubelet logs** :

```
Aug 06 12:09:35 aks-agentpool-12873730-1 kubelet[1756]: I0806 12:09:35.906299    1756 kubelet_node_status.go:491] Using Node Hostname from cloudprovider: "aks-agentpool-12873730-1"

Aug 06 12:09:29 aks-agentpool-12873730-1 kubelet[1756]: E0806 12:09:29.965335    1756 file.go:76] Unable to read manifest path "/etc/kubernetes/manifests": path does not exist, ignoring
Aug 06 12:09:26 aks-agentpool-12873730-1 kubelet[1756]: I0806 12:09:26.613745    1756 server.go:796] GET /stats/summary: (7.883039ms) 200 [[Ruby] 10.244.5.60:44376]
Aug 06 12:09:25 aks-agentpool-12873730-1 kubelet[1756]: I0806 12:09:25.781823    1756 kubelet_node_status.go:491] Using Node Hostname from cloudprovider: "aks-agentpool-12873730-1"
Aug 06 12:09:21 aks-agentpool-12873730-1 kubelet[1756]: E0806 12:09:21.092330    1756 pod_workers.go:186] Error syncing pod f13eb1ad-9970-11e8-8f58-0a58ac1f107d ("yocto-agent-10ssr-3x38j_jen
kins(f13eb1ad-9970-11e8-8f58-0a58ac1f107d)"), skipping: timeout expired waiting for volumes to attach or mount for pod "jenkins"/"yocto-agent-10ssr-3x38j". list of unmounted volumes=[output]
. list of unattached volumes=[output workspace-volume default-token-vx9mp]
Aug 06 12:09:21 aks-agentpool-12873730-1 kubelet[1756]: E0806 12:09:21.092261    1756 kubelet.go:1640] Unable to mount volumes for pod "yocto-agent-10ssr-3x38j_jenkins(f13eb1ad-9970-11e8-8f5
8-0a58ac1f107d)": timeout expired waiting for volumes to attach or mount for pod "jenkins"/"yocto-agent-10ssr-3x38j". list of unmounted volumes=[output]. list of unattached volumes=[output w
orkspace-volume default-token-vx9mp]; skipping pod
Aug 06 12:09:15 aks-agentpool-12873730-1 kubelet[1756]: I0806 12:09:15.747747    1756 kubelet_node_status.go:491] Using Node Hostname from cloudprovider: "aks-agentpool-12873730-1"
Aug 06 12:09:09 aks-agentpool-12873730-1 kubelet[1756]: E0806 12:09:09.965090    1756 file.go:76] Unable to read manifest path "/etc/kubernetes/manifests": path does not exist, ignoring
Aug 06 12:09:05 aks-agentpool-12873730-1 kubelet[1756]: I0806 12:09:05.710028    1756 kubelet_node_status.go:491] Using Node Hostname from cloudprovider: "aks-agentpool-12873730-1"
```

**Pod**
```
Name:                      yocto-agent-10ssr-sjcxn
Namespace:                 jenkins
Node:                      aks-agentpool-12873730-1/10.240.0.5
Start Time:                Mon, 06 Aug 2018 14:58:09 +0300
Labels:                    jenkins=slave
                           jenkins/yocto-agent=true
Annotations:               <none>
Status:                    Terminating (lasts 14m)
Termination Grace Period:  30s
IP:
Containers:
  slave:
    Container ID:
    Image:         *********/builder-yocto-slave:3
    Image ID:
    Port:          <none>
    Host Port:     <none>
    Command:
      cat
    State:          Terminated
      Exit Code:    0
      Started:      Mon, 01 Jan 0001 00:00:00 +0000
      Finished:     Mon, 01 Jan 0001 00:00:00 +0000
    Ready:          False
    Restart Count:  0
    Environment:    <none>
    Mounts:
      /home/jenkins from workspace-volume (rw)
      /output from output (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-vx9mp (ro)
  jnlp:
    Container ID:
    Image:          jenkins/jnlp-slave:alpine
    Image ID:
    Port:           <none>
    Host Port:      <none>
    State:          Terminated
      Exit Code:    0
      Started:      Mon, 01 Jan 0001 00:00:00 +0000
      Finished:     Mon, 01 Jan 0001 00:00:00 +0000
    Ready:          False
    Restart Count:  0
    Environment:
      JENKINS_SECRET:  **************
      JENKINS_NAME:    yocto-agent-10ssr-sjcxn
      JENKINS_URL:     ***************
      HOME:            /home/jenkins
    Mounts:
      /home/jenkins from workspace-volume (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-vx9mp (ro)
Conditions:
  Type           Status
  Initialized    True
  Ready          False
  PodScheduled   True
Volumes:
  output:
    Type:       PersistentVolumeClaim (a reference to a PersistentVolumeClaim in the same namespace)
    ClaimName:  yocto-premium-output
    ReadOnly:   false
  workspace-volume:
    Type:    EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:
  default-token-vx9mp:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  default-token-vx9mp
    Optional:    false
QoS Class:       BestEffort
Node-Selectors:  <none>
Tolerations:     node.kubernetes.io/not-ready:NoExecute for 300s
                 node.kubernetes.io/unreachable:NoExecute for 300s
Events:
  Type     Reason                  Age                From                               Message
  ----     ------                  ----               ----                               -------
  Normal   Scheduled               22m                default-scheduler                  Successfully assigned yocto-agent-10ssr-sjcxn to aks-agentpool-12873730-1
  Normal   SuccessfulMountVolume   22m                kubelet, aks-agentpool-12873730-1  MountVolume.SetUp succeeded for volume "workspace-volume"
  Normal   SuccessfulMountVolume   22m                kubelet, aks-agentpool-12873730-1  MountVolume.SetUp succeeded for volume "default-token-vx9mp"
  Normal   SuccessfulAttachVolume  21m                attachdetach-controller            AttachVolume.Attach succeeded for volume "pvc-f2ad9da3-90a6-11e8-a8f5-0a58ac1f149a"
  Warning  FailedMount             15m (x3 over 20m)  kubelet, aks-agentpool-12873730-1  Unable to mount volumes for pod "yocto-agent-10ssr-sjcxn_jenkins(fce1a7aa-996f-11e8-8f58-0a58ac1f107d)": timeout expired waiting for volumes to attach or mount for pod "jenkins"/"yocto-agent-10ssr-sjcxn". list of unmounted volumes=[output]. list of unattached volumes=[output workspace-volume default-token-vx9mp]

```

**PV / PVC Log**
```
Name:          yocto-premium-output
Namespace:     jenkins
StorageClass:  managed-premium
Status:        Bound
Volume:        pvc-f2ad9da3-90a6-11e8-a8f5-0a58ac1f149a
Labels:        <none>
Annotations:   pv.kubernetes.io/bind-completed=yes
               pv.kubernetes.io/bound-by-controller=yes
               volume.beta.kubernetes.io/storage-provisioner=kubernetes.io/azure-disk
Finalizers:    [kubernetes.io/pvc-protection]
Capacity:      100Gi
Access Modes:  RWO
Events:        <none>
```

**Environment**:
- Kubernetes version (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.6", GitCommit:"a21fdbd78dde8f5447f5f6c331f7eb6f80bd684e", GitTreeState:"clean", BuildDate:"2018-07-26T10:04:08Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}
```
- Cloud provider or hardware configuration:
   `Azure Kubernetes Service`

**Comment by andyzhangx** [comment]
Hi @hmenn , has it recovered now? It looks like mount timeout is due to long time disk attachment (expected time is less than 1 min). And what's your k8s region, then?

**Comment by hmenn** [comment]
Hi @andyzhangx , yes. It takes  around 13 minutes to reach first pipeline step. All resources are located in East US region. 

I'm also using dynamic persistent volume claims like this. 
```
kind: PersistentVolumeClaim
apiVersion: v1
metadata:
  name: SOME_NAME
  namespace: jenkins
spec:
  accessModes:
    - ReadWriteOnce
  storageClassName: managed-premium # Remove this line if you want hdd
  resources:
    requests:
      storage: 100Gi
```

P.S: I connected k8s node over ssh to see disk attachments. I can see content of disk on k8s node but k8s service not mounts to pod.

**Comment by andyzhangx** [comment]
@hmenn so after a few tries by the pod, the pod is in running state now?

**Comment by hmenn** [comment]
@andyzhangx I'm using this pod for building some projects. The pod die after finish pipeline. For each recreation, I'm waiting extra 10 minutes for disk attachments to pod.

**Comment by hmenn** [comment]
@andyzhangx  using eastus is not mandatory. If you have another advice for europe which is tested, I can create another resource group. And I will also test eastus2

**Comment by andyzhangx** [comment]
@hmenn attaching disk time cost is not stable, if your workload finishes quickly, I would suggest use azure file instead, the mount would be much quicker.

**Comment by hmenn** [comment]
@andyzhangx Azure file storage is case insensitive but I need case sensitive file system. So, Do I have to accept this timeout problem overhead? Actually 10 minutes acceptable for one pod but when we start another pod, it takes about two hours to mount second disk. I'm looking for another alternatives. 

https://docs.microsoft.com/en-us/azure/storage/common/storage-decide-blobs-files-disks

**Comment by andyzhangx** [comment]
> it takes about two hours to mount second disk
It's a bug, could you repro that stably? on which region?  @hmenn

**Comment by hmenn** [comment]
@andyzhangx I tested on eastus servers. But westeurope is also same.

Steps:
1. Start first job from jenkins(This jobs uses pvc1). Jenkins triggers kubernetes.
2. Kubernetes tries to start pod in node2 and gets mount timeout(for pvc1) error.
3. Start 2. Job from jenkins(This jobs uses pvc2). Kubernetes also deploys it in node2.
4. After 15 minutes first jobs starts and then finishes successfully. But second pod still getting timout error.
5. after about 2 hours waiting, second pod is ready to use.

The only error message from my kubelet is "Mount timeout".

**Comment by andyzhangx** [comment]
Could you run "top" command on node2 to check whether CPU load is high on that VM?
It should not wait for 2 hours.
if possible, to isolate this issue, could you run two deployments on same node and check total time cost? follow by: 
https://github.com/andyzhangx/demo/tree/master/linux/azuredisk/attach-stress-test/deployment

So you are using AKS 1.10.6 on eastus with node size: D4s_v3, right?

**Comment by hmenn** [comment]
Yes. Configs are right. I had test cpu loads when I faced with this problem first time. CPU load was less than 2(VM has 4vCPU so it's a good ratio) while it was waiting for job2.

These are drain results. For example, moving azuredisk2 takes about 8 minutes.
```
Successfully assigned deployment-azuredisk2-65f7457b86-kz82h to aks-agentpool-12873730-0
default-scheduler - 1 2018-08-09T07:50 UTC 2018-08-09T07:50 UTC
warning Multi-Attach error for volume "pvc-145efb4d-9ba4-11e8-9889-0a58ac1f003f" Volume is already used by pod(s) deployment-azuredisk2-65f7457b86-lmgqx attachdetach-controller - 1 2018-08-09T07:50 UTC 2018-08-09T07:50 UTC
MountVolume.SetUp succeeded for volume "default-token-lcgjv" kubelet aks-agentpool-12873730-0 - 1 2018-08-09T07:50 UTC 2018-08-09T07:50 UTC 
**_warning Unable to mount volumes for pod "deployment-azuredisk2-65f7457b86-kz82h_default(d4fac449-9ba8-11e8-9889-0a58ac1f003f)": timeout expired waiting for volumes to attach or mount for pod "default"/"deployment-azuredisk2-65f7457b86-kz82h". list of unmounted volumes=[azuredisk]. list of unattached volumes=[azuredisk default-token-lcgjv] kubelet aks-agentpool-12873730-0 - 3 2018-08-09T07:52 UTC 2018-08-09T07:56 UTC_**
AttachVolume.Attach succeeded for volume "pvc-145efb4d-9ba4-11e8-9889-0a58ac1f003f" attachdetach-controller - 1 2018-08-09T07:57 UTC 2018-08-09T07:57 UTC 
MountVolume.SetUp succeeded for volume "pvc-145efb4d-9ba4-11e8-9889-0a58ac1f003f" kubelet aks-agentpool-12873730-0 -1 2018-08-09T07:58 UTC 2018-08-09T07:58 UTC
pulling image "nginx" kubelet aks-agentpool-12873730-0 spec.containers{deployment-azuredisk2} 1 2018-08-09T07:58 UTC 2018-08-09T07:58 UTC
Successfully pulled image "nginx" kubelet aks-agentpool-12873730-0 spec.containers{deployment-azuredisk2} 1 2018-08-09T07:58 UTC 2018-08-09T07:58 UTC
Created container kubelet aks-agentpool-12873730-0 spec.containers{deployment-azuredisk2} 1 2018-08-09T07:58 UTC 2018-08-09T07:58 UTC 
Started container kubelet aks-agentpool-12873730-0 spec.containers{deployment-azuredisk2} 1
```

**Comment by andyzhangx** [comment]
@hmenn That's a little slow which it's also acceptable for you, right? (eastus2 region may be a little bit faster)

`AttachVolume.Attach succeeded` means attach disk completes and `MountVolume.SetUp succeeded` means disk recognization and format completes. So in your worst case(wait for 2 hours?) , could you provide same pod logs by `kubectl describe po POD-NAME`, I would like to know which step is sucking? And you may also go to azure portal to check whehther that disk is already attached to VM.

**Comment by andyzhangx** [comment]
It looks like your two pods yocto-agent-z0xws-cjl99 and yocto-agent-z0xws-lhn85 are using same PVC disk: yocto-premium-output, it's not allowed to do that, azure disk is RWO

**Comment by hmenn** [comment]
@andyzhangx I only started 1 pod. After too many fails it terminates old one and created new one. Can i increase timeout?
It does not give multiple attach error. If i start another pod manually, it will give that error.

**Comment by andyzhangx** [comment]
Hi @hmenn I tried my [deployment scripts](https://github.com/andyzhangx/demo/tree/master/linux/azuredisk/attach-stress-test/deployment) on AKS 1.10.6 on eastus with node size: D4s_v3, takes about 4 min for two pods with one azure disk each scheduled from one node to another, is there anything I am missing?
```
Events:
  Type     Reason                  Age   From                               Message
  ----     ------                  ----  ----                               -------
  Normal   Scheduled               15m   default-scheduler                  Successfully assigned deployment-azuredisk2-65f7457b86-77qqp to aks-nodepool1-35448466-1
  Warning  FailedAttachVolume      15m   attachdetach-controller            Multi-Attach error for volume "pvc-ad77dbae-9c65-11e8-8122-0a58ac1f0768" Volume is already used by pod(s) deployment-azuredisk2-65f7457b86-vstj7
  Normal   SuccessfulMountVolume   15m   kubelet, aks-nodepool1-35448466-1  MountVolume.SetUp succeeded for volume "default-token-995dq"
  Warning  FailedMount             13m   kubelet, aks-nodepool1-35448466-1  Unable to mount volumes for pod "deployment-azuredisk2-65f7457b86-77qqp_default(06f924de-9c6a-11e8-80cd-0a58ac1f0aac)": timeout expired waiting for volumes to attach or mount for pod "default"/"deployment-azuredisk2-65f7457b86-77qqp". list of unmounted volumes=[azuredisk]. list of unattached volumes=[azuredisk default-token-995dq]
  Normal   SuccessfulAttachVolume  12m   attachdetach-controller            AttachVolume.Attach succeeded for volume "pvc-ad77dbae-9c65-11e8-8122-0a58ac1f0768"
  Normal   SuccessfulMountVolume   11m   kubelet, aks-nodepool1-35448466-1  MountVolume.SetUp succeeded for volume "pvc-ad77dbae-9c65-11e8-8122-0a58ac1f0768"
  Normal   Pulling                 11m   kubelet, aks-nodepool1-35448466-1  pulling image "nginx"
  Normal   Pulled                  10m   kubelet, aks-nodepool1-35448466-1  Successfully pulled image "nginx"
  Normal   Created                 10m   kubelet, aks-nodepool1-35448466-1  Created container
  Normal   Started                 10m   kubelet, aks-nodepool1-35448466-1  Started container
```

**Comment by hmenn** [comment]
@andyzhangx No. Can you try to increase disk size to 100GB? My First deployment also done in 5-6 minutes. After writing some data to disk(~50GB), mount timeout error occurs. I think there is a problem about disk formatting and recognition when disk has data.

**Comment by andyzhangx** [comment]
@hmenn ok, I will try. Would you also run `sudo journalctl -u kubelet -l > kubelet.log` on agent node if there is timeout error? Thanks.

**Comment by andyzhangx** [comment]
I tried 100GB disk and write ~60GB data, drain two pods with one disk each would cost about 5 min and it works well. And your provided kubelet logs has issue in attach disk other that mount device.
Would you open the kubelet log level to 12, follow by, and retry, this will give you more detailed logs if it could step into mount device process:
```
edit yaml file under /etc/systemd/system/kubelet.service, change --v=2 value(e.g. change to --v=12) and then run

sudo vi /etc/systemd/system/kubelet.service
#edit 
sudo systemctl daemon-reload
sudo systemctl restart kubelet
```

**Comment by hmenn** [comment]
Hi @andyzhangx , I uploaded latest logs to drive because of huge size. 

https://drive.google.com/file/d/1q7oXAKGn5gs6mmLFyKquZEhS525-w_ms/view?usp=sharing

**Comment by hmenn** [comment]
@andyzhangx Did you look the logs? I'm still getting "Unable to mount volumes for pod" error. Exact 12 min to mount disk to pod on each run.

**Comment by andyzhangx** [comment]
@hmenn is it possible to share your exact config so that I could repro it on my env? The provided logs only show that the disk attaching to the agent is slow, I did not find any error when mounting disk device.

**Comment by hmenn** [comment]
@andyzhangx I think I found the problem.

On first template, I was using fsGroup because of some permission to write disk. My docker starts with nonroot user and this prevents to writing to pvc. So I had added fsGroup option. Today, I just removed and started pod again, it started in 1 minutes. I guess on each disk attach, its trying to change group id of all files which is in disk and this causing to time out error.

with fsGroup:
```
apiVersion: v1
kind: Pod
metadata:
    label: some-label-here
    name: some-name-here
spec:
  securityContext:
     fsGroup: 2000
  containers:
  - name: slave
    image: image:1
    command:
    - cat
    tty: true
    volumeMounts:
    - name: output
      mountPath: /output
  imagePullSecrets:
  - name: ps-secret
  volumes:
  - name: output
    persistentVolumeClaim:
      claimName: aosp-premium-output
```

without fsGroup:
```
apiVersion: v1
kind: Pod
metadata:
    label: some-label-here
    name: some-name-here
spec:
  containers:
  - name: slave
    image: image:1
    command:
    - cat
    tty: true
    volumeMounts:
    - name: output
      mountPath: /output
  imagePullSecrets:
  - name: ps-secret
  volumes:
  - name: output
    persistentVolumeClaim:
      claimName: aosp-premium-output
```

I test this templates a few times with same env. Adding fsGroups causes timeout error.

So, do we need to change default docker users? Is there another way to write data to disk from nonroot user?

**Comment by andyzhangx** [comment]
what about set fsgroup as 0:
```
apiVersion: v1
kind: Pod
metadata:
  name: security-context-demo
spec:
  securityContext:
    runAsUser: 0
    fsGroup: 0
```

**Comment by andyzhangx** [comment]
azureDisk use ext4 file system by default, mountOptions like [uid=x,gid=x] could not be set in mount time.

**Comment by hmenn** [comment]
I changed docker default users to root. Now, my workspace works fine. Thank you so much @andyzhangx . Do i need to open another bug for using fsGroup with huge file systems?

**Comment by andyzhangx** [comment]
@hmenn It's not necessary since it's a common issue for mounting ex4 file system, not related to k8s. I have recorded your issue in my doc. Would you close this issue? Thanks.

**Comment by hmenn** [comment]
I'm closing the issue.

**Comment by v-braun** [comment]
Lost so many hours on that! Thanks!
Just a question:
is it possible to increase the mountin timeout?
My pod restarts after 2m and needs a lot of restarts till it stay alive.
I think an increase of the timeout would help to improve the time because kubes chown can process till and and has not be re-executed multiple times.

**Comment by andyzhangx** [comment]
> Lost so many hours on that! Thanks!
> Just a question:
> is it possible to increase the mountin timeout?
> My pod restarts after 2m and needs a lot of restarts till it stay alive.
> I think an increase of the timeout would help to improve the time because kubes chown can process till and and has not be re-executed multiple times.

@v-braun for chown slowness issue, you could use following workaround:
Following setting will set pod run as root, make it accessible to any file:
```
apiVersion: v1
kind: Pod
metadata:
  name: security-context-demo
spec:
  securityContext:
    runAsUser: 0
    fsGroup: 0
```

 - another workaround is using `supplementalGroups`:
```
apiVersion: v1
kind: Pod
...
spec:
  containers:
  - name: ...
    volumeMounts:
    - name: xxx 
      mountPath: /usr/share/... 
  securityContext: 
    supplementalGroups: [xxx] 
  volumes:
```

**Comment by v-braun** [comment]
Excuse me if this is a stupid question but would ``runAsUser: 0`` not mean the process inside the container runs as root. Has it not an impact on security of the container?

The ``supplementalGroup`` sounds better from security perspective or?

**Comment by andyzhangx** [comment]
> Excuse me if this is a stupid question but would `runAsUser: 0` not mean the process inside the container runs as root. Has it not an impact on security of the container?
> 
> The `supplementalGroup` sounds better from security perspective or?

correct, `supplementalGroup` is safer.

**Comment by pantelis-karamolegkos** [comment]
We are experiencing a similar issue.

We are running `jenkins` on GKE 1.16.

We are using the official helm [chart](https://github.com/helm/charts/tree/master/stable/jenkins) to deploy it.

Very often, after a restart, the master pod fails to go into `Running` state because of failed mount error.

```
  Normal   Scheduled    11m                  default-scheduler                                Successfully assigned jenkins/jenkins-cd-cb8b56fb7-ssp2j to gke-environment-nodepool1-3d1a1f31-vw87
  Warning  FailedMount  20s (x5 over 9m23s)  kubelet, gke-environment-nodepool1-3d1a1f31-vw87  Unable to mount volumes for pod "jenkins-cd-cb8b56fb7-ssp2j_jenkins(412467g16-2bd5-48f5-b536-5de124c65dc80)": timeout expired waiting for volumes to attach or mount for pod "jenkins"/"jenkins-cd-cb8b56fb7-ssp2j". list of unmounted volumes=[jenkins-home]. list of unattached volumes=[plugins tmp jenkins-config secrets-dir jenkins-home jenkins-cd-token-83l7h]
```

The problem usually gets resolved by itself after a long time, usually (a lot more than) 10 minutes.

I am posting in this thread although from the discussion and if I am not mistaken it seems that the topic discussed relates more to the slave pods?

**Comment by rawon** [comment]
I had this problem (using Azure) too. Apparently this is because the service principal credential is expired. If you're on Azure try this: https://docs.microsoft.com/en-us/azure/aks/update-credentials#update-aks-cluster-with-new-service-principal-credentials

**Comment by huychau** [comment]
Fixed by update `securityContext` on Alibaba cloud, Kubernetes version 1.9.3:

```
securityContext:
    ...
    privileged: true
    runAsUser: 0
```

**Comment by pedrohdz** [comment]
OK..  I did some digging around to get a better understanding of the problem and figured I would post it here for others that run into similar issues with long volume mount times. Apologies in advance, I know little to no GoLang.

If I'm reading the Kubernetes code correctly, The Azure Managed Disk extension (code?) causing the delay appears to be a call to [SetVolumeOwnership()](https://github.com/kubernetes/kubernetes/blob/36bde432adefb1f6200889379ada974bb9f3c48d/pkg/volume/azuredd/azure_mounter.go#L166-L168).  Not sure if `azuredd` is the right place for Azure Managed Disk.

[SetVolumeOwnership()](https://github.com/kubernetes/kubernetes/blob/36bde432adefb1f6200889379ada974bb9f3c48d/pkg/volume/volume_linux.go#L43) traverses the the entire file system when `fsGroup` is set.  This can can be avoided if  `fsGroupChangePolicy` is set to `OnRootMismatch` in the pod security context.  This is assuming that the volume's root permissions match. 

So the more files on the filesystem, the longer it is going to take.  In our case we have over 2,000,000 files in on the volume on a light day, and why it would take anywhere from 25 to 40 minutes the pod to gain access to the volume.

To fix this, the choices are either using:
- `supplementalGroups` instead of `fsGroup` in the pod security context.
- set `fsGroup`, which is not recommended for security reasons.
- if you have to use `fsGroup`, try setting `fsGroupChangePolicy` to `OnRootMismatch`, *I have not tested this out*, but if I'm reading the code correctly this should work as well.

Some quick looking around in `kubernetes/kubernetes/pkg/volume` it seems like `SetVolumeOwnership()` is used in all the volume types **except** for: `azure_file`, `cephfs`, `csimigration`, `glusterfs`, `hostpath`, `nfs`, `quobyte`.  That list should not be considered definitive though.

**Comment by guillermobandres** [comment]
> 
> 
> OK.. I did some digging around to get a better understanding of the problem and figured I would post it here for others that run into similar issues with long volume mount times. Apologies in advance, I know little to no GoLang.
> 
> If I'm reading the Kubernetes code correctly, The Azure Managed Disk extension (code?) causing the delay appears to be a call to [SetVolumeOwnership()](https://github.com/kubernetes/kubernetes/blob/36bde432adefb1f6200889379ada974bb9f3c48d/pkg/volume/azuredd/azure_mounter.go#L166-L168). Not sure if `azuredd` is the right place for Azure Managed Disk.
> 
> [SetVolumeOwnership()](https://github.com/kubernetes/kubernetes/blob/36bde432adefb1f6200889379ada974bb9f3c48d/pkg/volume/volume_linux.go#L43) traverses the the entire file system when `fsGroup` is set. This can can be avoided if `fsGroupChangePolicy` is set to `OnRootMismatch` in the pod security context. This is assuming that the volume's root permissions match.
> 
> So the more files on the filesystem, the longer it is going to take. In our case we have over 2,000,000 files in on the volume on a light day, and why it would take anywhere from 25 to 40 minutes the pod to gain access to the volume.
> 
> To fix this, the choices are either using:
> 
>     * `supplementalGroups` instead of `fsGroup` in the pod security context.
> 
>     * set `fsGroup`, which is not recommended for security reasons.
> 
>     * if you have to use `fsGroup`, try setting `fsGroupChangePolicy` to `OnRootMismatch`, _I have not tested this out_, but if I'm reading the code correctly this should work as well.
> 
> 
> Some quick looking around in `kubernetes/kubernetes/pkg/volume` it seems like `SetVolumeOwnership()` is used in all the volume types **except** for: `azure_file`, `cephfs`, `csimigration`, `glusterfs`, `hostpath`, `nfs`, `quobyte`. That list should not be considered definitive though.

Hi @pedrohdz I'm having the same problem in my AKS with a log volume with 50Gb of size. I tried to use supplementalGroups instead of fsGroup but it still taken a lot of time to start the container. The version of k8s is 1.17.1 and this is my security config:

      securityContext:
        runAsGroup: 10001
        runAsNonRoot: true
        runAsUser: 10001
        supplementalGroups:
          - 10001
The storage class is Azure managed disk, is there something wrong??

Thank you!

**Comment by pedrohdz** [comment]
Hey @guillermobandres!
I don't know to be honest.  What I can say is that we are using a `securityContext` along the lines of:

```yaml
  securityContext:
    runAsNonRoot: true
    runAsUser: 10001
    supplementalGroups:
      - 10001
```

Note that we are not using `runAsGroup`.  Maybe that is triggering the `chown` as well?

If that does not work, you can try running it as root like @andyzhangx  mentioned in: https://github.com/kubernetes/kubernetes/issues/67014#issuecomment-589915496, though I would not recommend running as root in production.

**Comment by andreas-eberle** [comment]
> 
> So the more files on the filesystem, the longer it is going to take. In our case we have over 2,000,000 files in on the volume on a light day, and why it would take anywhere from 25 to 40 minutes the pod to gain access to the volume.
> 
> To fix this, the choices are either using:
> 
>     * `supplementalGroups` instead of `fsGroup` in the pod security context.
> 
>     * set `fsGroup`, which is not recommended for security reasons.
> 
>     * if you have to use `fsGroup`, try setting `fsGroupChangePolicy` to `OnRootMismatch`, _I have not tested this out_, but if I'm reading the code correctly this should work as well.
> 

@pedrohdz: You are right. They introduced the `fsGroupChangePolicy` in [Kubernetes 1.18 as an alpha feature](https://v1-18.docs.kubernetes.io/docs/tasks/configure-pod-container/security-context/#configure-volume-permission-and-ownership-change-policy-for-pods) and with 1.20. it became beta.

**Comment by guillermobandres** [comment]
> 
> 
> Hey @guillermobandres!
> I don't know to be honest. What I can say is that we are using a `securityContext` along the lines of:
> 
> ```yaml
>   securityContext:
>     runAsNonRoot: true
>     runAsUser: 10001
>     supplementalGroups:
>       - 10001
> ```
> 
> Note that we are not using `runAsGroup`. Maybe that is triggering the `chown` as well?
> 
> If that does not work, you can try running it as root like @andyzhangx mentioned in: [#67014 (comment)](https://github.com/kubernetes/kubernetes/issues/67014#issuecomment-589915496), though I would not recommend running as root in production.

Hi pedrohdz I tried with the security context as you said:

  securityContext:
    runAsNonRoot: true
    runAsUser: 10001
    supplementalGroups:
      - 10001

But the pod still taking a lot of time to start. Now the k8s cluster is on 1.18 version. Did it work on your cluster with this configuration??

**Comment by pedrohdz** [comment]
Hey @guillermobandres,

Don't know then.  You might want to try experimenting with mounting an empty file system, or one with fewer files, to see if that makes a difference?  Does the file system have a lot of files?  It could potentially be something else as well.  In our particular case it was too many files.

**Comment by guillermobandres** [comment]
> 
> 
> Hey @guillermobandres,
> 
> Don't know then. You might want to try experimenting with mounting an empty file system, or one with fewer files, to see if that makes a difference? Does the file system have a lot of files? It could potentially be something else as well. In our particular case it was too many files.

Hi @pedrohdz, yes, the disk have a lot files, this deployment is Loki, that create a lot of small files. I'll try with root permission just to confirm it works as we expect. Today it was 1h 10m to mount the volume...

**Comment by gsusI** [comment]
Seems like the issue was introduced into Google Cloud Platform on the 2020-04-20 (🌈 420). [Source](https://serverfault.com/questions/1060925/gke-fails-to-mount-volumes-to-deployments-pods-timed-out-waiting-for-the-condit/1060936#1060936).

It's difficult to establish causality here, but I started facing this issue after updating to from `1.17.17-gke.3700` to `1.18.16-gke.502`, and got fixed when updating from `1.18.16-gke.502` to 1.19.8-gke.1600.

**Comment by liurd** [comment]
Hi,  @pedrohdz 

Even remove fsGroup in yaml, 
`
securityContext:
  runAsUser: 1000
  runAsNonRoot: true
  supplementalGroups:
    - 1000
`
While after installation, the default 
`  securityContext:
    fsGroup: 1
    runAsNonRoot: true
    runAsUser: 1000
    seccompProfile:
      type: RuntimeDefault
    supplementalGroups:
    - 1000
`
What's your result? where is fsGroup: 1 from?
